ONYX-11477: Update Ruby conventions page (#94)
cyberark/community: Conjur/conventions/ruby-style-guide.md

as we refine and hone our Ruby knowledge base.

## Style Baseline
- Follow this official guidelines of the Ruby language https://docs.ruby-lang.org/en/master/doc/syntax/methods_rdoc.html
  
- Methods whose purpose is to produce a side effect (change a file, make a network call, raise an error, etc) should have verb phrase names. 

  ```ruby
  def call
    validate_and_decode_token
    get_jwt_identity_from_request
    validate_host_has_access_to_webservice
    validate_origin
    validate_restrictions
    audit_success
    new_token
  rescue => e
    audit_failure(e)
    raise e
  end
  ```

  

> - Functions that return an object should have the name of the returned object so these functions should not have verb phrase names.

Reviewer: "Functions" -- Let's use "Methods" since that is the proper ruby terminology here.

Also add the sentence:

In some cases, these rules may conflict when a side-effectful method also returns a value.   Here the "verb phrase" rule should win since indicating the side effect is more important.

- For low-level style decisions about formatting, etc, we follow the ruby style. We advice to use rubocop to auto correct offence in the files you change before submitting to PR.

> - Dont add "name stutter" to the names of files/classes/functions/members if its already in the name of the package because its obvious . For example inside authn_jwt package there is no need to add the "name stutter" jwt to whats inside because its obvious.

Reviewer: Small update for correct terminology:

Avoid "name stutter".  That is, don't repeat a name from the module path in a class, and don't repeat a name from the class in a method.   For example:
```ruby
# bad
class MyCar
  def start_car_engine
    # ...
  end
end

# good
class MyCar
  def start_engine
    # ...
  end
end

# bad
module Authenticators
  class OktaAuthenticator
    #...
  end
end

# good
module Authenticators
  class Okta
    #...
  end
end

- Dont add "name stutter" to the names of files/classes/functions/members if its already in the name of the package because its obvious . For example inside authn_jwt package there is no need to add the "name stutter" jwt to whats inside because its obvious.

- In command class we should split methods called in `call` function into 2 parts - one with an explicit name so that the command class 'call' function tells the story, and one with the returned object name for memoization.

  - Bad:

    ```ruby
    def get_jwt_identity_from_request
        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
        @jwt_identity ||= @jwt_configuration.jwt_identity
        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
    end
    ```

  - Good:

    ```ruby
    def get_jwt_identity_from_request
        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
        jwt_identity
        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
    end
    
    def jwt_identity
        @jwt_identity ||= @jwt_configuration.jwt_identity
    end
    ```

> - In constuctor of classes seperate input handling and dependency handling with empty line

Reviewer 2: spelling: `seperate` => `separate`

> - In command class impliment methods in the order they are called

Reviewer 2: spelling: `impliment` => `implement`

    create_identity_provider: Authentication::AuthnJwt::IdentityProviders::CreateIdentityProvider.new
  )
  ```

- Use memoization when you need to run "hard operation" more than one and you sure that the output is the same (DB queries, API requests, etc..) . Use the '||=' to do it. The name of the function in memoization and the name of the memoaized object should be the same.

  ```ruby
  def identity_role
    @identity_role ||= @role_class.by_login(
      jwt_identity,
      account: account
    )
  end
  ```

- When you want to run code that assignes value to class member in the first time it runs and afterwards it just returns it you need to check if you already run it, and if so to return the class member.

  - Check if its first for  boolean object:

    ```ruby
    def identity_available?
      return @identity_available if [true, false].include?(@identity_available)
    
      @identity_available = username_exists?
    end
    ```

  - Check if its first for non boolean object

    ```ruby
    def x
      return @x if @x
    
      @x = calculate_x()
    end
    ```

> - Put all instance methods that are not being called under private keyword.

Reviewer 2: This is already covered by the rubocop style guide that we link to from the convention README (see: https://github.com/rubocop/ruby-style-guide#leverage-access-modifiers-eg-private-and-protected). Granted they don't go into as much detail by showing good / bad cases but it's probably simple enough that we don't need to.

Author: I will remove this section

    validate_host_has_access_to_webservice
    validate_origin
    validate_restrictions
    audit_success
    new_token
  rescue => e
    audit_failure(e)
    raise e
  end
  ```

  

- Methods that return an object should have the name of the returned object so these functions should not have verb phrase names.
   In some cases, these rules may conflict when a side-effectful method also returns a value. Here the "verb phrase" rule should win since indicating the side effect is more important.

- For low-level style decisions about formatting, etc, we follow the ruby style. We advice to use rubocop to auto correct offence in the files you change before submitting to PR.

- Avoid "name stutter". That is, don't repeat a name from the module path in a class, and don't repeat a name from the class in a method. For example:

   ```ruby
   # bad
   class MyCar
     def start_car_engine
       # ...
     end
   end
   
   # good
   class MyCar
     def start_engine
       # ...
     end
   end
   
   # bad
   module Authenticators
     class OktaAuthenticator
       #...
     end
   end
   
   # good
   module Authenticators
     class Okta
       #...
     end
   end
   ```

> - In command class we should split methods called in `call` function into 2 parts - one with an explicit name so that the command class 'call' function tells the story, and one with the returned object name for memoization.

Reviewer 3: Can you please give some more detail on why the second option is better? It looks like we're just adding an extra method call, which feels like it clouds to story. I always prefer concise but understandable over verbose.

Author: Removed this rule following this comment
https://github.com/cyberark/community/pull/94#discussion_r691518622

  - Bad:

    ```ruby
    def get_jwt_identity_from_request
        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
        @jwt_identity ||= @jwt_configuration.jwt_identity
        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
    end
    ```

  - Good:

    ```ruby
    def get_jwt_identity_from_request
        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
        jwt_identity
        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
    end
    
    def jwt_identity
        @jwt_identity ||= @jwt_configuration.jwt_identity
    end

>     ```

Author: Ok i will remove it from document

- In constuctor of classes separate input handling and dependency handling with empty line

- In command class implement methods in the order they are called

  ```ruby
  def call
    func_a
    func_b
    func_c
  end
  
  private
  
  def func_a
    ...
  end
  
  def func_b
    ...
  end
  
  def sub_func_b
    ...
  end
  
  def func_c
    ...
    end
  ```

- In functions with many parameters put each parameter in seperate line in the declaration. In this case end each param with ":" so the caller will need to specify name of variable for each param.

  ```
  def method(
    param1:
    param2:
    param3:
  )
  ```

  

> - When overriding a code climate violation with an inline exception (this isn't possible for all code climate  plugins but is for many), add the reason for the exception on the line above the special "ignore" directive

Reviewer 3: I'd expect this to be super uncommon. I've used overrides in the case of improved readability. The default complexity and method length settings can be problematic as well. Coding standards push people to write higher quality, easier to understand code. If there are a lot of warnings, it's more likely an issue of code not being as mature as we'd like it to be.

Author: Regarding this i think its the job of the PR reviewer to check each warning in the PR and decide if he accepts it or wants the code to change. WDYT

Author: I will remove this section and when we agree it will be done in another pr

Reviewer: @jvanderhoof I don't take this rule as saying it's ok to do lots of overrides (though perhaps we should clarify that it _isn't_), but only "_If you _have to_ do an override for a legitimate reason, then you need to document the reason as a comment"

   https://github.com/cyberark/conjur/blob/5fe6b22f225a1f2cf0be90d90f184c74c150be5a/cucumber/policy/features/support/rest_client_wrapper.rb#L12

## Design Baseline

- On any new class ask yourself if it should be a class or command class. Command Classes coordinate the execution of a multi-step process.

  https://github.com/cyberark/conjur/blob/master/app/domain/README.md#the-functional-object-oriented-style

  

> - No need for interface classes

Reviewer 3: Ruby is duck typed, so if you need to define an interface, try handling it with tests.

> - *Never* use inheritance or mixins.

Reviewer 3: Instead, use composition.

Author: By composition you mean function calling another function?

> - Dependencies for classes should be passed as default values in constructor

Reviewer 3: ```suggestion
- Dependencies for classes should be passed in the constructor
```
I'm seeing a misunderstanding of dependency injection which is easy to miss with the Command class.  All the command class does is provide a nice DSL over vanilla Ruby:

```ruby
module Authenticators
  class Okta
    def initialize(logger:, configuration:)
      @logger = logger
      @configuration = configuration
    end

    def call(request:)
      ....
    end
  end
end
```

vs

```ruby
module Authenticators
  class Okta
    extend CommandClass::Include

    command_class(
      dependencies: {logger:, configuration:},
      inputs: %i[request]
    ) do

      def call(request:)
        ....
      end
    end
  end
end  
```

In the top example, the initialization arguments depend on duck-typing (the interface of the dependency matters).  @jonahx's Command class reduces the amount of boilerplate code, but (IMO), reduces the transparency of exactly how an injected dependency is injected (and whether a dependency should be injected into the `initialize` or `call` method).

I'd love to see us moving injection up the stack, so:

```ruby 
Authentication::AuthnOidc::UpdateInputWithUsernameFromIdToken.new.(
  authenticator_input: authenticator_input
)
```

becomes:

```ruby
Authentication::AuthnOidc::UpdateInputWithUsernameFromIdToken.new(
  fetch_authenticator_secrets: Authentication::Util::FetchAuthenticatorSecrets.new(
    fetch_secrets: ::Conjur::FetchRequiredSecrets.new(
      resource_class: ::Resource
    )
  ),
  validate_account_exists: ::Authentication::Security::ValidateAccountExists.new(
    role_class: ::Role
  ),
  verify_and_decode_token: ::Authentication::OAuth::VerifyAndDecodeToken.new(
    fetch_provider_keys: ::Util::ConcurrencyLimitedCache.new(
      ::Util::RateLimitedCache.new(
        FetchProviderKeys.new(
          logger: logger,
          discover_identity_provider: DiscoverIdentityProvider.new(
            logger: logger,
            open_id_discovery_service: OpenIDConnect::Discovery::Provider::Config
          )
        ),
        refreshes_per_interval: 10,
        rate_limit_interval: 300, # 300 seconds (every 5 mins)
        logger: logger
      ),
      max_concurrent_requests: 3, # TODO: Should be dynamic calculation
      logger: logger
    ),
    verify_and_decode_token: ::Authentication::Jwt::VerifyAndDecodeToken.new(
      jwt_decoder: JWT,
      logger: logger
    ),
    logger: logger
  ),
  logger: logger
).call(
  authenticator_input: authenticator_input
)
```

Although the second option is substantially more verbose, it's very clear what the dependencies are.  We can clearly see the OIDC authenticator uses the Conjur `Role` and `Resource`, something that was not obvious when default values were used int the initialization block.

Author: Im not sure if its really readable . 
logger: logger repeates it self so many times in a single place its little bit strange to me.

Reviewer 3: @tzheleznyak, exactly the point of being explicit. In the current implementation, the method call looks very simple and straightforward.  In truth, it's anything but. In reality, the `UpdateInputWithUsernameFromIdToken` initializer has 23 dependencies (in the entire tree). 

If this complexity is unsatisfactory, we should deal with the root cause of the complexity. Maybe we move Policy loading outside the authenticator, maybe we refactor the two implementations of `VerifyAndDecodeToken` to use a single class rather than two seperate (`Jwt` and  `OAuth`). The test is the time it takes a new engineer to come up to speed, the seemingly "simple" top level call that hide the complexity. 

IMO, the number of security issues stemming from authenticators can be traced to their complexity.  Simple to understand code is far less likely to have bugs.

  ```ruby
  def initialize(
    authenticator_input:,
    logger: Rails.logger,
    validate_resource_restrictions_class: Authentication::ResourceRestrictions::ValidateResourceRestrictions,
    extract_token_from_credentials: Authentication::AuthnJwt::InputValidation::ExtractTokenFromCredentials.new,
    create_identity_provider: Authentication::AuthnJwt::IdentityProviders::CreateIdentityProvider.new
  )
  ```

> - Use memoization when you need to run "hard operation" more than one and you sure that the output is the same (DB queries, API requests, etc..) . Use the '||=' to do it. The name of the function in memoization and the name of the memoaized object should be the same.

Reviewer 3: ```suggestion
- Use memoization when you need to run "expensive operation" more than once and the output is the same (DB queries, API requests, etc..) . Use the '||=' to do it. The name of the function in memoization and the name of the memoized object should be the same.
```

Mostly interested in swapping "hard" for "expensive". Memoization increases memory consumption but is a small price to pay to avoid the hit of re-querying an API for the same data.

  ```ruby
  def identity_role
    @identity_role ||= @role_class.by_login(
      jwt_identity,
      account: account
    )
  end
  ```

> - When you want to run code that assignes value to class member in the first time it runs and afterwards it just returns it you need to check if you already run it, and if so to return the class member.

Reviewer 3: I don't agree this guidance. I've seen this create a lot of confusion in the flow of code. 

In OOP, it's entirely appropriate for state to be maintained inside an object. In a functional programming, all methods really should be stateless (95% of the time - the exception is memoization).  Instead, just use memoization:

```ruby
def user_name
  expensive_api_call[:name]
end

# returns a hash
def expensive_api_call(client:, query:)
  @expensive_api_call ||= client.fetch(query)
end
```

I also want to clarify what the `@var ||= action` actually does. It performs a truthy test the value of the variable (`@var`).  If `@var` is set, it is returned.  If `@var` is not set, `action` is assigned to `@var` and returned (as the result of the last line of a method is returned by default).

Author: Not sure if i understand. If i have hard boolean operation that i want to memoaize call to it. And the return value is false. In case i don't check the memoaized value is false it will run again .

```
def boolean_hard_operation?
    # Perform api calls
    # Perform db queries
end
```

Reviewer 3: @tzheleznyak, in that case, I'd recommend caching the responses of the above API and DB calls:

```ruby 
def boolean_hard_operation?
  # Perform api calls
  api_call 
  # Perform db queries
  db_call
end

def api_call
  @api_call ||= make_expensive_api_call
end

def db_call
  @db_call ||= make_expensive_db_call
end
```

The boolean logic is still calculated, but we're not focused on memoizing the boolean value (it's not the boolean calculation that is slow, it's the api and db calls required to by the boolean calculation that is slow).

  - Check if its first for  boolean object:

    ```ruby
    def identity_available?
      return @identity_available if [true, false].include?(@identity_available)
    
      @identity_available = username_exists?
    end
    ```

  - Check if its first for non boolean object

    ```ruby
    def x
      return @x if @x
    
      @x = calculate_x()
    end

>     ```

Reviewer: I think the second bullet makes it confusing.  Instead, I would keep this focused on just being an explanation for the special case of a boolean value.  Suggested edit:

- For the special case when the value being memoized is boolean, the above method won't work.  That's because both `nil` and `false` evaluate to false, so the `||=` pattern can't distinguish between an uninitialized value and and already initialized value that happens to be `false`.  For this case we suggest the following explicit check:

    ```ruby
    def identity_available?
      return @identity_available if [true, false].include?(@identity_available)
    
      @identity_available = username_exists?
    end
    ```

> - A command class should never call new on anything besides a value object. If it needs to do that, the “already -new’d” thing should have been passed in to begin with.  Related to this: value object classes are not dependencies and can be referred to directly. So they will never need to be passed in.

Reviewer 3: Can you please provide an example here?  I think the point is more related to value objects than command classes.  If so, the distinction is really important.  Value objects can really simplify how data is passed and transformed within a system. Value objects should be immutable, and are better suited then Command classes for some of the "is it set yet" workflows mentioned above.

  Articles 
- **Learning Ruby** ([The Hard Way](https://learnrubythehardway.org/book/)) - A simple book meant to
  get you started in programming with Ruby
- **Command Class** ([Github Public Page](https://github.com/jonahx/command_class)) - Read more about CommandClass design pattern implementations

>    that frequently used in Conjur.

Reviewer: I would also include a link to this video, which provides more context for these ideas:

https://www.youtube.com/watch?v=7qnsRejCyEQ&t=1387s
